### ISSUE TITLE
design: Tariff based invoices - Cases

### ISSUE THREAD

**Issue by sfount** [third-culture-software/bhima#1770]
**Problem**
A number of installation sites that we have visited are moving to using a tariff based system of billing. Instead of patients being responsible for covering the cost of their health care, they are instead assigned a tariff. This is a set amount that will pay no matter what happens with their progression through the hospital. 

This is possible to model in BHIMA however without billing for individual invoice items there will be no way of knowing how much treating patients is actually costing the hospital. The only two data points available would be the cost that the hospital should make from tariffs and how much they actually do make in cash. The cost of healthcare would be be unknown. 

**Cases Concept** 
After a patient has been registered they can create a 'Case'. Creating a case will invoice the patient based on tariff. This case can then be used to cover individual invoices. 

Invoices can *optionally* reference a case, if an invoice references a case the debt will immediately be balanced by the hospital. 

**Requirements** 
* Enterprise must have an expense account that can be used to immediately balance transactions invoiced under cases 
* Invoice creation must have a feature that creates a voucher to immediately balance any invoice created and filed under a case 
* An invoice must be able to reference a case UUID
* Cases will have to be an entity - they can only be created by billing a patient for a tariff (referenced in the case)
* Tariffs will have to be an entity - potentially with related inventory item, these will be used to create invoices when creating a case. Patient's will be given an invoice for a tariff 
* We will probably want a flag that will only show invoices created referencing a case or invoices that are not referencing cases

**Result** 
* When an invoice is made with a referenced case, as soon as the invoice is made and the receipt is downloaded it will show that the invoice has already been paid by the hospital
* When an invoice is made that doesn't reference a case it will behave exactly as BHIMA does right now

**Comment by DedrickEnc** [comment]
@sfount
Thanks to create this issue, my question is : 
- Is it possible to use both case system and normal system?
- By reading this issue, I have not see where you are talking about changes which must happen on cash module, should we create a separate issue for that?

**Comment by sfount** [comment]
@DedrickEnc 

1. I would suggest that cases are completely optional and should not be required. We could add an option in enterprise settings so that if an enterprise has no use for the optional case fields they will not be displayed. 

2. The cash module should have no changes at all given this design. Cash should be paid towards invoices in exactly the same way.

**Comment by sfount** [comment]
I have also proposed a much simpler solution that would not require as much infrastructure to build - it would not track individual cases but it would allow enterprises to optionally assign any invoice to be covered fully by the enterprises 'case' account. 

*Quote* 

I do have a suggestion for an intermediate solution that would not take as long to develop and would be less complex for the hospital: 

* The Enterprise has a configuration option to set an expense account that will cover patient invoices
* Every invoice has the option to be marked as 'Covered by a case'. This would just be a simple check box that is either checked or not. 
* When an invoice is created that is 'Covered by a case' it will immediately be balanced to 0 against the Enterprises expense account. The patient is still billed it is just immediately paid by the hospital. 

This involves very little infrastructure on our part. It then provides to options for billing. 

1. 'Covered by a case' is not checked. The patient is invoiced (for example for the original tariff) and owes that money to the hospital. 
2. 'Covered by a case' is checked. The patient is invoiced (for example for medicines) and the invoice is instantly paid by the hospital, no money is owed by the patient.

**Comment by DedrickEnc** [comment]
Yes but technically how the cash module will load bills? because even if the invoice is covered by the case, the patient should pay.
If the cash will not change, that means the patient invoice module will create two invoices (one balanced and other not balanced)?

**Comment by sfount** [comment]
The cash modules should load bills in the same way. 

If the invoice is covered by a 'Case' the balance will be paid immediately by the hospital - because the balance will always be 0 it will never show up in the cash module. 

If the invoice is not covered by a 'Case' the balance will be > 0 and it will show up in the cash module.

**Comment by jniles** [comment]
If we are going to implement this, I would be more comfortable going all the way in and supporting all the infrastructure required to do Tarif-based invoicing against a case.

This was a really detailed issue, and we can probably start to work on it anytime.  It will expand our potential installation sites greatly.
